File: app/src/main/java/com/aas/moviecatalog/fragment/MovieFragment.java
package com.aas.moviecatalog.fragment;


import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.aas.moviecatalog.detail.DetailMovieActivity;
import com.aas.moviecatalog.R;
import com.aas.moviecatalog.adapter.MovieAdapter;
import com.aas.moviecatalog.model.Movie;
import com.aas.moviecatalog.viewmodel.ListMovieViewModel;

import java.util.ArrayList;


/**
 * A simple {@link Fragment} subclass.
 */
public class MovieFragment extends Fragment implements MovieAdapter.MovieListener {

    private MovieAdapter movieAdapter;
    private RecyclerView rvMovie;
    private ArrayList<Movie> listMovie;
    private ProgressBar progressBar;
    private ListMovieViewModel movieViewModel;

    private void showLoading(boolean loading){
        if (loading){
            progressBar.setVisibility(View.VISIBLE);
        }
        else {
            progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_movie, container, false);

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        rvMovie = view.findViewById(R.id.rvMovie);
        progressBar = view.findViewById(R.id.pbMovie);
        movieAdapter = new MovieAdapter(getContext(), this);

        movieViewModel = ViewModelProviders.of(this)
                .get(ListMovieViewModel.class);
        movieViewModel.getListMovie().observe(this, getListMovie);

        //rvMovie.setHasFixedSize(true);
        rvMovie.setLayoutManager(new LinearLayoutManager(getContext()));
        rvMovie.setAdapter(movieAdapter);

        showLoading(true);
        movieViewModel.setListMovies();
    }

    private Observer<ArrayList<Movie>> getListMovie = new Observer<ArrayList<Movie>>() {
        @Override
        public void onChanged(ArrayList<Movie> movies) {
            if (movies != null){
                listMovie = movies;
                movieAdapter.setListMovie(movies);
            }
            showLoading(false);
        }
    };

    @Override
    public void onMovieItemClick(int position) {
        Intent intent = new Intent(getContext(), DetailMovieActivity.class);
        intent.putExtra(DetailMovieActivity.EXTRA_MOVIE, listMovie.get(position));
        startActivity(intent);
    }
}

File: app/src/main/java/com/aas/moviecatalog/movie/MovieFragment.kt
package com.aas.moviecatalog.movie


import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.aas.moviecatalog.MainActivity
import com.aas.moviecatalog.R
import com.aas.moviecatalog.api.ApiRepository
import com.aas.moviecatalog.detail.DetailActivity
import com.aas.moviecatalog.model.Movie
import org.jetbrains.anko.*
import kotlinx.android.synthetic.main.fragment_movie.*

/**
 * A simple [Fragment] subclass.
 */
class MovieFragment : Fragment(), MovieInterface {
    private lateinit var mPresenter: MoviePresenter
    private lateinit var mAdapter: MovieAdapter
    private var movies = mutableListOf<Movie>()
    private lateinit var viewModel: MovieViewModel
    private lateinit var rvMovie: RecyclerView

    override fun showLoading() {
        pbMovie.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        pbMovie.visibility = View.GONE
    }

    override fun movieData(movie: MovieResponseModel) {
        viewModel.setMovie(movie)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_movie, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        rvMovie = view.findViewById(R.id.rvMovie)
        super.onViewCreated(view, savedInstanceState)
    }

    @Suppress("DEPRECATION")
    override fun onActivityCreated(savedInstanceState: Bundle?) {

        viewModel = ViewModelProviders.of(this).get(MovieViewModel::class.java)
        viewModel.getMovie().observe(this, getMovie)

        val service = ApiRepository.create()
        mPresenter = MoviePresenter(this, service)

        rvMovie.addItemDecoration(
            DividerItemDecoration(
                rvMovie.context,
                DividerItemDecoration.VERTICAL
            )
        )
        rvMovie.layoutManager = LinearLayoutManager(requireContext())
        mAdapter = MovieAdapter(requireContext(), movies) {
            val intent = Intent(context, DetailActivity::class.java)
            intent.putExtra(MainActivity.DATA_EXTRA, it.id)
            intent.putExtra(MainActivity.TYPE, MainActivity.MOVIE)
            startActivity(intent)
        }
        rvMovie.adapter = mAdapter
        if (savedInstanceState == null) {
            mPresenter.loadMovie()
        } else {
            hideLoading()
        }

        super.onActivityCreated(savedInstanceState)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putString(MainActivity.INSTANCE, MovieFragment::class.java.simpleName)
        super.onSaveInstanceState(outState)
    }

    private val getMovie = Observer<MovieResponseModel> {
        if (it != null) {
            mAdapter.setMovie(it.results)
        }
    }
}
